File: dsa_leetcode_beginner/0417-pacific-atlantic-water-flow/0417-pacific-atlantic-water-flow.cpp
#include <vector>

using namespace std;

class Solution {
public:
    vector<vector<int>> pacificAtlantic(vector<vector<int>>& heights) {
        vector<vector<int>> result;
        if (heights.empty()) return result;

        int m = heights.size(), n = heights[0].size();
        vector<vector<bool>> pacificReachable(m, vector<bool>(n, false));
        vector<vector<bool>> atlanticReachable(m, vector<bool>(n, false));

        for (int i = 0; i < m; ++i) {
            dfs(heights, pacificReachable, i, 0);
            dfs(heights, atlanticReachable, i, n - 1);
        }

        for (int j = 0; j < n; ++j) {
            dfs(heights, pacificReachable, 0, j);
            dfs(heights, atlanticReachable, m - 1, j);
        }

        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                if (pacificReachable[i][j] && atlanticReachable[i][j]) {
                    result.push_back({i, j});
                }
            }
        }

        return result;
    }

private:
    void dfs(vector<vector<int>>& heights, vector<vector<bool>>& reachable, int row, int col) {
        reachable[row][col] = true;
        vector<pair<int, int>> directions = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};

        for (auto& direction : directions) {
            int newRow = row + direction.first;
            int newCol = col + direction.second;
            if (newRow >= 0 && newRow < heights.size() && newCol >= 0 && newCol < heights[0].size() &&
                !reachable[newRow][newCol] && heights[newRow][newCol] >= heights[row][col]) {
                dfs(heights, reachable, newRow, newCol);
            }
        }
    }
};

File: dsa_leetcode_beginner/0417-pacific-atlantic-water-flow/0417-pacific-atlantic-water-flow.js
var pacificAtlantic = function(heights) {
    if (!heights || heights.length === 0) {
        return [];
    }

    const m = heights.length, n = heights[0].length;
    const pacificReachable = Array.from({ length: m }, () => Array(n).fill(false));
    const atlanticReachable = Array.from({ length: m }, () => Array(n).fill(false));

    const dfs = (row, col, reachable) => {
        reachable[row][col] = true;
        const directions = [[1, 0], [-1, 0], [0, 1], [0, -1]];

        for (let [dr, dc] of directions) {
            const newRow = row + dr, newCol = col + dc;
            if (newRow >= 0 && newRow < m && newCol >= 0 && newCol < n &&
                !reachable[newRow][newCol] && heights[newRow][newCol] >= heights[row][col]) {
                dfs(newRow, newCol, reachable);
            }
        }
    };

    for (let i = 0; i < m; i++) {
        dfs(i, 0, pacificReachable);
        dfs(i, n - 1, atlanticReachable);
    }

    for (let j = 0; j < n; j++) {
        dfs(0, j, pacificReachable);
        dfs(m - 1, j, atlanticReachable);
    }

    const result = [];
    for (let i = 0; i < m; i++) {
        for (let j = 0; j < n; j++) {
            if (pacificReachable[i][j] && atlanticReachable[i][j]) {
                result.push([i, j]);
            }
        }
    }

    return result;
};
